src/main/java/org/cru/cas/client/integration/CopyCasAttributesToHeadersFilter.java

```diff
@@ -0,0 +1,251 @@
+package org.cru.cas.client.integration;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.function.BinaryOperator;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletRequestWrapper;
+import org.jasig.cas.client.authentication.AttributePrincipal;
+
+/**
+ * "Copies" CAS attributes from the request's Principal to a set of configured headers.
+ * The mapping is defined by the filter init-param {@code attributeMapping}.
+ *
+ *
+ * @author Matt Drees
+ */
+public class CopyCasAttributesToHeadersFilter implements Filter {
+
+    static final String ATTRIBUTE_MAPPING_PARAMETER = "attributeMapping";
+    private static final String LINE_SEPARATOR = "=";
+    private static final String CAS_PREFIX = "CAS_";
+
+    private AttributeMapping attributeMapping;
+
+    interface AttributeMapping {
+        public boolean headerIsMapped(String headerName);
+
+        public String getAttributeNameForHeader(String headerName);
+
+        public Set<String> getHeaderNamesForAttribute(String attributeName);
+    }
+
+    @Override
+    public void init(final FilterConfig filterConfig) throws ServletException {
+        String mapping = filterConfig.getInitParameter(ATTRIBUTE_MAPPING_PARAMETER);
+        if (mapping != null && !mapping.isEmpty()) {
+            this.attributeMapping = parseAttributeMapping(mapping);
+        } else {
+            this.attributeMapping = new AttributeMapping() {
```

Comment:
How about devoting this section to its own method?

Reply:
yeah, I think I was going to return to this but forgot